Scanner file archiver

mnonnenmacher opened the pull request:
Add the feature to archive part of the scanned files in a file storage. By default only license files are archived, but this can be changed in the configuration. See the commit messages for details.

sschuberth commented on model/src/main/kotlin/config/FileStorageConfiguration.kt:
```diff
@@ -36,4 +37,20 @@ data class FileStorageConfiguration(
+    /**
+     * Create a [FileStorage] based on this configuration.
+     */
+    fun create(): FileStorage {
```
Nit: Just "create" in the context of this class sounds like to create a file storage *configuration*, maybe `createStorage`?

mnonnenmacher replied:
Done.

sschuberth commented on model/src/main/kotlin/config/FileSaverConfiguration.kt:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Create a [FileSaver] based on this configuration.
+     */
+    fun create() = FileSaver(patterns, storage.create())
```
Again, maybe `createSaver`?

mnonnenmacher replied:
Done.

sschuberth commented on scanner/src/main/kotlin/LocalScanner.kt:
```diff
@@ -40,15 +40,9 @@ import com.here.ort.model.ScannerDetails
-import com.here.ort.utils.log
-import com.here.ort.utils.safeMkdirs
-import com.here.ort.utils.showStackTrace
+import com.here.ort.utils.*
```
Nit: No wildcard import here.

mnonnenmacher replied:
Done.

sschuberth commented on scanner/src/main/kotlin/LocalScanner.kt:
```diff
@@ -61,12 +55,32 @@ import kotlinx.coroutines.asCoroutineDispatcher
 import kotlinx.coroutines.async
 import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.withContext
+import java.security.MessageDigest
```
Nit: Blank line before this one.

mnonnenmacher replied:
Done.

sschuberth commented on model/src/test/kotlin/config/OrtConfigurationTest.kt:
```diff
@@ -55,7 +67,7 @@ class OrtConfigurationTest : WordSpec({
                         backend.localFileStorage.let { localFileStorage ->
                             localFileStorage shouldNotBe null
-                            localFileStorage!!.directory shouldBe File("~/.ort/scanner")
+                            localFileStorage!!.directory shouldBe File("~/.ort/scanner/results")
```
Shouldn't this be part of https://github.com/heremaps/oss-review-toolkit/pull/1899?

mnonnenmacher replied:
Added to #1899, I will rebase this PR later.

sschuberth commented on model/src/main/resources/reference.conf:
```diff
@@ -13,7 +22,7 @@ ort {
 
         localFileStorage {
-          directory = ~/.ort/scanner
+          directory = ~/.ort/scanner/results
```
Shouldn't this be part of https://github.com/heremaps/oss-review-toolkit/pull/1899?

mnonnenmacher replied:
Added to #1899, I will rebase this PR later.

sschuberth commented on scanner/src/main/kotlin/LocalScanner.kt:
```diff
@@ -61,12 +55,32 @@ import kotlinx.coroutines.asCoroutineDispatcher
+    companion object {
+        val DEFAULT_ARCHIVE_DIR by lazy { getUserOrtDirectory().resolve("scanner/archive") }
+
+        val DEFAULT_ARCHIVE_PATTERNS = listOf(
```
Hmm, this basically duplicates the list in `FileMatcher`. Looks to be like we should actually keep the global `LICENSE_FILE_NAMES` that was removed in a0d12dfa91b37650b8b8b78d277d690757266ed2 and use it here as well as in `FileMatcher`.

mnonnenmacher replied:
It's not exactly the same list, but I agree it makes sense to have the common part in a single place. I would address this in a separate PR if that is ok for you.

sschuberth replied:
Yes, that's fine!

sschuberth commented on scanner/src/main/kotlin/LocalScanner.kt:
```diff
@@ -300,13 +314,26 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
+        val provenanceHash = MessageDigest.getInstance("SHA-1").digest(provenanceBytes).toHexString()
+        val path = "${id.toPath()}/$provenanceHash"
+
+        archiver.save(directory, path)
```
Just asking myself aloud: Would this work as expected for consecutive scans of the same package where only the pattern for which files to archive changed? I.e. would any new matching files be added to the directory of the previous scan, but not any not-anymore-matching files be removed? I *do* believe that *is* the case from reading the code, but I want to double-check with you.

mnonnenmacher replied:
Depends on what you expect :)
But yes, currently it would add any new files but not delete anything, because the `FileSaver` can only (over)write files.
However, in a new scan of the same package the timestamp of the provenance would be different resulting in a different provenance hash, so a new directory would be used for archiving the files. I considered ignoring the timestamp when building the provenance hash, because while in theory the scanned files for the same provenance could have changed, in reality this is highly unlikely as the provenance contains either the SCM revision or the source artifact hash. I would like to keep it like that for now to be on the safe side, and maybe change it later, but nonetheless, what do you think about this?

sschuberth replied:
Yeah, I believe we can leave it like this for now until we eventually see a concrete need to fine-tune the hash generation.

But for a brief moment I was wondering whether we should use something completely different (i.e. not derived from the provenance) as the hash, like the SPDX package verification code for *all* files in the directory (not only the archived ones), as you can't deduce the provenance from the has anyway, and as that would have the advantage of not requiring additional storage for new package versions where the license files did not actually change.

What do you think?

mnonnenmacher replied:
We need to look up archived files in other components like the notice reporter based on the provenance of the scan result, so currently only the provenance hash works for this. But the SPDX verification hash is one of the next tasks I will be working on and it will be added to the scan result model, so once it is there we could consider switching to this instead.

sschuberth replied:
> But the SPDX verification hash is one of the next tasks I will be working on

Just as a reminder, we already have the (so far unused) `calculatePackageVerificationCode` functions in `SpdxUtils.kt`.

mnonnenmacher replied:
Nice, that will come in handy.

mnonnenmacher replied:
I made a PR for the SPDX verification code (#1903), what do you think, shall we use this instead for the path? I'm uncertain which is better.

sschuberth replied:
That depends on what we decide about https://github.com/heremaps/oss-review-toolkit/pull/1903#discussion_r338413577. IMO it makes only sense to consider the package verification code here if it refers to all files in the package (like you've implemented it for now), not only to the scanned files. But the latter also has its merit.

sschuberth commented on utils/src/main/kotlin/storage/HttpFileStorage.kt:
```diff
@@ -40,7 +40,7 @@ class HttpFileStorage(
      * The URL to store files at.
      */
-    private val url: String,
+    val url: String,
```
I couldn't find a place where this is publicly used now, is it?

mnonnenmacher replied:
It's used in `ScanResultsStorage:71` in a log message.

sschuberth commented on model/src/main/kotlin/config/FileSaverConfiguration.kt:
```diff
@@ -0,0 +1,40 @@
+import com.here.ort.utils.storage.FileSaver
+import com.here.ort.utils.storage.FileStorage
+
+data class FileSaverConfiguration(
```
For completeness, and to align with the other configuration classes, could we also get some class documentation here?

mnonnenmacher replied:
Done.

mnonnenmacher commented:
@sschuberth : FYI, since your review I have changed this PR to archive files in a ZIP file before upload. The main reason is that it is otherwise hard to download the archived files without knowing their names, but it also has performance benefits. Therefore I have renamed `FileSaver` to `FileArchiver` again, because that name now makes sense.

sschuberth commented on utils/src/main/kotlin/ArchiveUtils.kt:
```diff
@@ -182,8 +182,9 @@ fun File.unpack7Zip(targetDirectory: File) {
  * Pack the file into a ZIP [targetFile] using [Deflater.BEST_COMPRESSION]. If the file is a directory its content is
  * recursively added to the archive. Only regular files are added, e.g. symbolic links or directories are skipped. If
  * a [prefix] is specified, it is added to the file names in the ZIP file.
+ * If not all files shall be added to the archive a [filter] can be provided.
```
Is the renaming of `FileSaver` really supposed to be in this commit?

mnonnenmacher replied:
No, fixed.

sschuberth commented on utils/src/main/kotlin/ArchiveUtils.kt:
```diff
@@ -184,8 +184,13 @@ fun File.unpack7Zip(targetDirectory: File) {
+    overwrite: Boolean = false,
+    filter: (Path) -> Boolean = { true }
+) {
+    require(overwrite || !targetFile.exists()) {
```
Nit: Don't you think swapping the conditions ready a bit more natural?

mnonnenmacher replied:
I find the current order natural. It also does the cheaper check first.

sschuberth replied:
Ok, doing the cheaper check first makes sense.

sschuberth commented on utils/src/main/kotlin/FileMatcher.kt:
```diff
@@ -28,9 +28,11 @@ import java.nio.file.Paths
-     * The list of file names to consider for matching.
+     * The list of [glob patterns][1] to consider for matching.
+     *
+     * [1]: https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String)
```
In `Configuration.md` we link https://docs.oracle.com/javase/tutorial/essential/io/fileOps.html#glob instead, should we use it here, too? If not, we should probably still not use a Java 12 link, because we actually use Java 8 as a compile target.

mnonnenmacher replied:
Done.
